### PR TITLE
Feature/delete exp improvements

### PR DESCRIPTION
**Pull request by mckayward**
# Notes on these chages:

## Start to move API call logging and error handling to Client classes
Similar to changes made for the DataClient in #21. Just cleaning up calls related to deleting experiments as the code is fresh in my mind.

## Handling 404s for queued and running experiments
The DELETE experiment API endpoint will return a 404 for an experiement that doesn't exist, but also for queued and running experiments. For this reason, the error message that the ExperimentClient logs for a 404 includes information about all three of those possibilities, and is a bit wordy.  If you want to keep the check for 'queued' and 'running' states in the cli and clean up the log message, I can do that. Maybe a different HTTP status for queued/running experiments (422 maybe?) would be good in the future. But for now, this works just fine.

### CODE REVIEW

File: floyd/client/experiment.py [hunk 1]
```diff
@@ -40,6 +42,19 @@ def create(self, experiment_request):
...
+        try:
+            self.request("DELETE",
+                         "{}{}".format(self.url, id))
+            floyd_logger.info("Experiment {}: Deleted".format(id))
+            return True
+        except NotFoundException as e:
+            floyd_logger.info(
+                    ("Experiment {}: ERROR! A deletable experiment with this "
+                     "id was not found. Make sure you have the correct id and "
+                     "that the experiment is not "
```
Comment (narenst) [inline floyd/client/experiment.py]: It is unfortunate that the server returns 404 if the experiment is queued or running. It should ideally return 400.

File: floyd/cli/experiment.py [hunk 2]
```diff
@@ -164,26 +165,27 @@ def delete(ids, yes):
...
 
-        if not yes:
-            if not click.confirm("Delete Run: {}?".format(experiment.name), abort=False, default=False):
-                floyd_logger.info("Experiment {}: Skipped.".format(experiment.name))
-                continue
+        task_instance_id = get_module_task_instance_id(experiment.task_instances)
+        task_instance = TaskInstanceClient().get(task_instance_id) if task_instance_id else None
 
         if task_instance and task_instance.module_id:
             ModuleClient().delete(task_instance.module_id)
```
Comment (narenst) [inline floyd/cli/experiment.py]: It seems like ModuleClient.delete could fail but the delete operations will continue. It should be added to the failures at least so that the user can take appropriate action to delete it later.

File: floyd/cli/experiment.py [hunk 3]
```diff
@@ -164,26 +165,27 @@ def delete(ids, yes):
...
     for id in ids:
         experiment = ExperimentClient().get(id)
-        task_instance_id = get_module_task_instance_id(experiment.task_instances)
-        task_instance = TaskInstanceClient().get(task_instance_id) if task_instance_id else None
-
+        if not experiment:
+            failures = True
+            continue
 
-        if experiment.state in ["queued", "running"]:
```
Comment (narenst) [inline floyd/cli/experiment.py]: Can we keep this check? Without this the cli will delete the module even when the experiment is running which is not ideal.

File: floyd/cli/experiment.py (hunk 2 above)
Comment (mckayward) [inline floyd/cli/experiment.py]: Yeah I think adding this to the failures is a good idea. However, the cli currently doesn't offer a way to delete a module directly by its id, so we may wish to implement that later depending on how we resolve all the comments in this PR.

File: floyd/cli/experiment.py (hunk 3 above)
Comment (mckayward) [inline floyd/cli/experiment.py]: Yeah good catch. If we just attempt to delete the experiment first, that would also solve the problem, and allow us to decouple the cli from knowing why an experiment might fail to delete. I also added a test for this situation. What do you think of this approach?

It seems impossible to totally remove the risk of having either the experiment and the module be deleted and have the other fail, though, since they are separate calls to the API and either request could fail for a number of different reasons. (Maybe this is something that the API itself should take care of if we never have a reason to delete only one of them).

One problem we have if one fails and the other doesn't is that the cli currently doesn't offer an interface to delete a module directly.

Comment (narenst) [inline floyd/cli/experiment.py]: Agreed, I'm working on the backend to delete the module as part of the delete experiment operation. After I merge that change, I will remove the module delete step from here.

In the meantime, I like the approach of deleting the experiment first.

File: floyd/client/experiment.py (hunk 1 above)
Comment (narenst) [inline floyd/client/experiment.py]: I'm also changing the error message to return 400 if the experiment is queued or running. I can update the cli as part of the next release.

Comment (narenst) [inline floyd/client/experiment.py]: That change is tracked in #28
